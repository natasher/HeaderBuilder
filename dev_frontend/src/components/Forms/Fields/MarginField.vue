<template>
  <div class = "row">

    <div class = "col col-left">
      <span class = "title">{{ ( as === 'grid' ) ? "gridMargin" : fieldName }}</span>
    </div>

    <div class = "col col-right">
      <div class = "mfn-field mfnf-margin">

        <div class = "container" >
          <input
            v-model     = "valueTop"
            class       = "top"
            :name       = "wpId+'Top'"
            placeholder = "px"
            type        = "text" />

          <input
            v-model     = "valueRight"
            class       = "right"
            :name       = "wpId+'Right'"
            placeholder = "px"
            type        = "text" />

          <input
            v-model     = "valueBottom"
            class       = "bottom"
            :name       = "wpId+'Bottom'"
            placeholder = "px"
            type        = "text" />

          <input
            v-model     = "valueLeft"
            class       = "left"
            :name       = "wpId+'Left'"
            placeholder = "px"
            type        = "text" />
        </div>

      </div>
    </div>

  </div>
</template>

<script>
import { mapGetters, mapActions } from 'vuex'

export default {

  name: 'MarginField',

  props: {
    /**
     * The name of the field.
     * Displayed on the left column.
     */
    fieldName: {
      type    : String,
      required: true,
    },
    /**
     * Name of the field used by WordPress internals.
     */
    wpId: {
      type   : String,
      require: true,
    },
    /**
     * Prop determinates type of Margin field,
     * possible values: **""**, **row**, **grid**.
     */
    as: {
      type    : String,
      required: false,
      default : '',
    },
    /**
     * When `as` prop is set to row, `row` indicates which one.
     * Then margin values will be set to given row.
     */
    row: {
      type    : String,
      required: false,
      default : '',
    },
  },

  methods: {
    ...mapActions( 'fields', [ 'setModalFieldValue' ]),
  },

  computed: {
    ...mapGetters( 'fields', [ 'getMarginFieldValue' ]),

    valueTop: {
      get() {
        return this.getMarginFieldValue(
          this.wpId,
          this.as,
          this.row,
          'top'
        )
      },
      set( value ) {
        const val = {
          name    : this.wpId,
          position: 'top',
          value   : value,
          as      : this.as  || '',
          row     : this.row || '',
        }

        this.setModalFieldValue( val )
      },
    },

    valueRight: {
      get() {
        return this.getMarginFieldValue(
          this.wpId,
          this.as,
          this.row,
          'right'
        )
      },
      set(value) {
        const val = {
          name    : this.wpId,
          position: 'right',
          value   : value,
          as      : this.as  || '',
          row     : this.row || '',
        }

        this.setModalFieldValue( val )
      },
    },

    valueBottom: {
      get() {
        return this.getMarginFieldValue(
          this.wpId,
          this.as,
          this.row,
          'bottom'
        )
      },
      set( value ) {
        const val = {
          name    : this.wpId,
          position: 'bottom',
          value   : value,
          as      : this.as  || '',
          row     : this.row || '',
        }

        this.setModalFieldValue( val )
      },
    },

    valueLeft: {
      get() {
        return this.getMarginFieldValue(
          this.wpId,
          this.as,
          this.row,
          'left'
        )
      },
      set(value) {
        const val = {
          name    : this.wpId,
          position: 'left',
          value   : value,
          as      : this.as  || '',
          row     : this.row || '',
        }

        this.setModalFieldValue( val )
      },
    },

  },
}
</script>
