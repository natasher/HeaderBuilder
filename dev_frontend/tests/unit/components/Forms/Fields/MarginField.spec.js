import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex          from 'vuex'
import flushPromises from 'flush-promises'
import merge         from 'lodash.merge'
import MarginField   from '../../../../../src/components/Forms/Fields/MarginField.vue'

const localVue = createLocalVue()
localVue.use(Vuex)

const val = 'test value'

function createStore(overrides) {
  const defaultStoreConfig = {
    modules: {
      fields: {
        namespaced: true,
        actions: {
          setModalFieldValue: jest.fn()
        },
        getters: {
          getMarginFieldValue: jest.fn().mockReturnValue(
            jest.fn().mockReturnValue( val )
          ),
        }
      }
    }
  }

  return new Vuex.Store(
    merge( defaultStoreConfig, overrides )
  )
}

function createWrapper(overrides) {
  const defaultMountingOptions = {
    localVue,
    store: createStore(),
    propsData: {
      fieldName: 'Test Field',
      wpId     : 'testField',
    }
  }

  return shallowMount( MarginField, merge( defaultMountingOptions, overrides ))
}

describe('TextField.vue', () => {

  describe('snapshots:', () => {

    test('should match snapshot if located in item form', () => {
      const wrapper = createWrapper()

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match snapshot if located in row form', () => {
      const wrapper = createWrapper({
        propsData: {
          as : 'row',
          row: 'testRow',
        }
      })

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match snapshot if located in grid form', () => {
      const wrapper = createWrapper({
        propsData: {
          as: 'grid',
        }
      })

      expect( wrapper.element ).toMatchSnapshot()
    })

  })

  describe('top value:', () => {

    test('should get value from item form state if `as` prop is empty', async () => {
      const innerGetter         = jest.fn().mockReturnValue()
      const getMarginFieldValue = jest.fn().mockReturnValue( innerGetter )
      const store = createStore({
        modules: {
          fields: {
            getters: {
              getMarginFieldValue,
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })
      await flushPromises()

      expect(innerGetter).toBeCalledWith(
        wrapper.vm.wpId,
        '',
        '',
        'top'
      )
    })

    test('should get value from row state if `as` props is `row`', async () => {
      const innerGetter         = jest.fn().mockReturnValue()
      const getMarginFieldValue = jest.fn().mockReturnValue( innerGetter )
      const store = createStore({
        modules: {
          fields: {
            getters: {
              getMarginFieldValue,
            }
          }
        }
      })
      const propsData = {
          as : 'row',
          row: 'testRow',
        }
      const wrapper = createWrapper({
        store,
        propsData,
      })
      await flushPromises()

      expect(innerGetter).toBeCalledWith(
        wrapper.vm.wpId,
        propsData.as,
        propsData.row,
        'top'
      )
    })

    test('should get value from row state if `as` props is `grid`', async () => {
      const innerGetter         = jest.fn().mockReturnValue()
      const getMarginFieldValue = jest.fn().mockReturnValue( innerGetter )
      const store = createStore({
        modules: {
          fields: {
            getters: {
              getMarginFieldValue,
            }
          }
        }
      })
      const propsData = {
          as : 'grid',
        }
      const wrapper = createWrapper({
        store,
        propsData,
      })
      await flushPromises()

      expect(innerGetter).toBeCalledWith(
        wrapper.vm.wpId,
        propsData.as,
        '',
        'top'
      )
    })

    test('should set value if filled (item form ver.)', async () => {
      const setModalFieldValue = jest.fn()
      const store = createStore({
        modules: {
          fields: {
            actions: {
              setModalFieldValue,
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })

      const input = wrapper.find( `input[name = "${ wrapper.vm.wpId }Top"]` )
      input.trigger( 'input' )
      await flushPromises()

      expect(setModalFieldValue).toBeCalledWith(
        expect.anything(),
        expect.objectContaining({
          name    : wrapper.vm.wpId,
          position: 'top',
          value   : val,
          as      : '',
          row     : '',
        }),
        void 0
      )
    })

    test('should set value if filled (row ver.)', async () => {
      const setModalFieldValue = jest.fn()
      const store = createStore({
        modules: {
          fields: {
            actions: {
              setModalFieldValue,
            }
          }
        }
      })
      const propsData = {
        as : 'row',
        row: 'testRow',
      }
      const wrapper = createWrapper({
        store,
        propsData,
      })

      const input = wrapper.find( `input[name = "${ wrapper.vm.wpId }Top"]` )
      input.trigger( 'input' )
      await flushPromises()

      expect(setModalFieldValue).toBeCalledWith(
        expect.anything(),
        expect.objectContaining({
          name    : wrapper.vm.wpId,
          position: 'top',
          value   : val,
          as      : propsData.as,
          row     : propsData.row,
        }),
        void 0
      )
    })

    test('should set value if filled (grid ver.)', async () => {
      const setModalFieldValue = jest.fn()
      const store = createStore({
        modules: {
          fields: {
            actions: {
              setModalFieldValue,
            }
          }
        }
      })
      const propsData = {
        as : 'grid',
      }
      const wrapper = createWrapper({
        store,
        propsData,
      })

      const input = wrapper.find( `input[name = "${ wrapper.vm.wpId }Top"]` )
      input.trigger( 'input' )
      await flushPromises()

      expect(setModalFieldValue).toBeCalledWith(
        expect.anything(),
        expect.objectContaining({
          name    : wrapper.vm.wpId,
          position: 'top',
          value   : val,
          as      : propsData.as,
          row     : '',
        }),
        void 0
      )
    })

  })

  describe('right value:', () => {

    test('should get value from item form state if `as` prop is empty', async () => {
      const innerGetter         = jest.fn().mockReturnValue()
      const getMarginFieldValue = jest.fn().mockReturnValue( innerGetter )
      const store = createStore({
        modules: {
          fields: {
            getters: {
              getMarginFieldValue,
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })
      await flushPromises()

      expect(innerGetter).toBeCalledWith(
        wrapper.vm.wpId,
        '',
        '',
        'right'
      )
    })

    test('should get value from row state if `as` props is `row`', async () => {
      const innerGetter         = jest.fn().mockReturnValue()
      const getMarginFieldValue = jest.fn().mockReturnValue( innerGetter )
      const store = createStore({
        modules: {
          fields: {
            getters: {
              getMarginFieldValue,
            }
          }
        }
      })
      const propsData = {
          as : 'row',
          row: 'testRow',
        }
      const wrapper = createWrapper({
        store,
        propsData,
      })
      await flushPromises()

      expect(innerGetter).toBeCalledWith(
        wrapper.vm.wpId,
        propsData.as,
        propsData.row,
        'right'
      )
    })

    test('should get value from row state if `as` props is `grid`', async () => {
      const innerGetter         = jest.fn().mockReturnValue()
      const getMarginFieldValue = jest.fn().mockReturnValue( innerGetter )
      const store = createStore({
        modules: {
          fields: {
            getters: {
              getMarginFieldValue,
            }
          }
        }
      })
      const propsData = {
          as : 'grid',
        }
      const wrapper = createWrapper({
        store,
        propsData,
      })
      await flushPromises()

      expect(innerGetter).toBeCalledWith(
        wrapper.vm.wpId,
        propsData.as,
        '',
        'right'
      )
    })

    test('should set value if filled (item form ver.)', async () => {
      const setModalFieldValue = jest.fn()
      const store = createStore({
        modules: {
          fields: {
            actions: {
              setModalFieldValue,
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })

      const input = wrapper.find( `input[name = "${ wrapper.vm.wpId }Right"]` )
      input.trigger( 'input' )
      await flushPromises()

      expect(setModalFieldValue).toBeCalledWith(
        expect.anything(),
        expect.objectContaining({
          name    : wrapper.vm.wpId,
          position: 'right',
          value   : val,
          as      : '',
          row     : '',
        }),
        void 0
      )
    })

    test('should set value if filled (row ver.)', async () => {
      const setModalFieldValue = jest.fn()
      const store = createStore({
        modules: {
          fields: {
            actions: {
              setModalFieldValue,
            }
          }
        }
      })
      const propsData = {
        as : 'row',
        row: 'testRow',
      }
      const wrapper = createWrapper({
        store,
        propsData,
      })

      const input = wrapper.find( `input[name = "${ wrapper.vm.wpId }Right"]` )
      input.trigger( 'input' )
      await flushPromises()

      expect(setModalFieldValue).toBeCalledWith(
        expect.anything(),
        expect.objectContaining({
          name    : wrapper.vm.wpId,
          position: 'right',
          value   : val,
          as      : propsData.as,
          row     : propsData.row,
        }),
        void 0
      )
    })

    test('should set value if filled (grid ver.)', async () => {
      const setModalFieldValue = jest.fn()
      const store = createStore({
        modules: {
          fields: {
            actions: {
              setModalFieldValue,
            }
          }
        }
      })
      const propsData = {
        as : 'grid',
      }
      const wrapper = createWrapper({
        store,
        propsData,
      })

      const input = wrapper.find( `input[name = "${ wrapper.vm.wpId }Right"]` )
      input.trigger( 'input' )
      await flushPromises()

      expect(setModalFieldValue).toBeCalledWith(
        expect.anything(),
        expect.objectContaining({
          name    : wrapper.vm.wpId,
          position: 'right',
          value   : val,
          as      : propsData.as,
          row     : '',
        }),
        void 0
      )
    })

  })

  describe('bottom value:', () => {

    test('should get value from item form state if `as` prop is empty', async () => {
      const innerGetter         = jest.fn().mockReturnValue()
      const getMarginFieldValue = jest.fn().mockReturnValue( innerGetter )
      const store = createStore({
        modules: {
          fields: {
            getters: {
              getMarginFieldValue,
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })
      await flushPromises()

      expect(innerGetter).toBeCalledWith(
        wrapper.vm.wpId,
        '',
        '',
        'bottom'
      )
    })

    test('should get value from row state if `as` props is `row`', async () => {
      const innerGetter         = jest.fn().mockReturnValue()
      const getMarginFieldValue = jest.fn().mockReturnValue( innerGetter )
      const store = createStore({
        modules: {
          fields: {
            getters: {
              getMarginFieldValue,
            }
          }
        }
      })
      const propsData = {
          as : 'row',
          row: 'testRow',
        }
      const wrapper = createWrapper({
        store,
        propsData,
      })
      await flushPromises()

      expect(innerGetter).toBeCalledWith(
        wrapper.vm.wpId,
        propsData.as,
        propsData.row,
        'bottom'
      )
    })

    test('should get value from row state if `as` props is `grid`', async () => {
      const innerGetter         = jest.fn().mockReturnValue()
      const getMarginFieldValue = jest.fn().mockReturnValue( innerGetter )
      const store = createStore({
        modules: {
          fields: {
            getters: {
              getMarginFieldValue,
            }
          }
        }
      })
      const propsData = {
          as : 'grid',
        }
      const wrapper = createWrapper({
        store,
        propsData,
      })
      await flushPromises()

      expect(innerGetter).toBeCalledWith(
        wrapper.vm.wpId,
        propsData.as,
        '',
        'bottom'
      )
    })

    test('should set value if filled (item form ver.)', async () => {
      const setModalFieldValue = jest.fn()
      const store = createStore({
        modules: {
          fields: {
            actions: {
              setModalFieldValue,
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })

      const input = wrapper.find( `input[name = "${ wrapper.vm.wpId }Bottom"]` )
      input.trigger( 'input' )
      await flushPromises()

      expect(setModalFieldValue).toBeCalledWith(
        expect.anything(),
        expect.objectContaining({
          name    : wrapper.vm.wpId,
          position: 'bottom',
          value   : val,
          as      : '',
          row     : '',
        }),
        void 0
      )
    })

    test('should set value if filled (row ver.)', async () => {
      const setModalFieldValue = jest.fn()
      const store = createStore({
        modules: {
          fields: {
            actions: {
              setModalFieldValue,
            }
          }
        }
      })
      const propsData = {
        as : 'row',
        row: 'testRow',
      }
      const wrapper = createWrapper({
        store,
        propsData,
      })

      const input = wrapper.find( `input[name = "${ wrapper.vm.wpId }Bottom"]` )
      input.trigger( 'input' )
      await flushPromises()

      expect(setModalFieldValue).toBeCalledWith(
        expect.anything(),
        expect.objectContaining({
          name    : wrapper.vm.wpId,
          position: 'bottom',
          value   : val,
          as      : propsData.as,
          row     : propsData.row,
        }),
        void 0
      )
    })

    test('should set value if filled (grid ver.)', async () => {
      const setModalFieldValue = jest.fn()
      const store = createStore({
        modules: {
          fields: {
            actions: {
              setModalFieldValue,
            }
          }
        }
      })
      const propsData = {
        as : 'grid',
      }
      const wrapper = createWrapper({
        store,
        propsData,
      })

      const input = wrapper.find( `input[name = "${ wrapper.vm.wpId }Bottom"]` )
      input.trigger( 'input' )
      await flushPromises()

      expect(setModalFieldValue).toBeCalledWith(
        expect.anything(),
        expect.objectContaining({
          name    : wrapper.vm.wpId,
          position: 'bottom',
          value   : val,
          as      : propsData.as,
          row     : '',
        }),
        void 0
      )
    })

  })

  describe('left value:', () => {

    test('should get value from item form state if `as` prop is empty', async () => {
      const innerGetter         = jest.fn().mockReturnValue()
      const getMarginFieldValue = jest.fn().mockReturnValue( innerGetter )
      const store = createStore({
        modules: {
          fields: {
            getters: {
              getMarginFieldValue,
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })
      await flushPromises()

      expect(innerGetter).toBeCalledWith(
        wrapper.vm.wpId,
        '',
        '',
        'left'
      )
    })

    test('should get value from row state if `as` props is `row`', async () => {
      const innerGetter         = jest.fn().mockReturnValue()
      const getMarginFieldValue = jest.fn().mockReturnValue( innerGetter )
      const store = createStore({
        modules: {
          fields: {
            getters: {
              getMarginFieldValue,
            }
          }
        }
      })
      const propsData = {
          as : 'row',
          row: 'testRow',
        }
      const wrapper = createWrapper({
        store,
        propsData,
      })
      await flushPromises()

      expect(innerGetter).toBeCalledWith(
        wrapper.vm.wpId,
        propsData.as,
        propsData.row,
        'left'
      )
    })

    test('should get value from row state if `as` props is `grid`', async () => {
      const innerGetter         = jest.fn().mockReturnValue()
      const getMarginFieldValue = jest.fn().mockReturnValue( innerGetter )
      const store = createStore({
        modules: {
          fields: {
            getters: {
              getMarginFieldValue,
            }
          }
        }
      })
      const propsData = {
          as : 'grid',
        }
      const wrapper = createWrapper({
        store,
        propsData,
      })
      await flushPromises()

      expect(innerGetter).toBeCalledWith(
        wrapper.vm.wpId,
        propsData.as,
        '',
        'left'
      )
    })

    test('should set value if filled (item form ver.)', async () => {
      const setModalFieldValue = jest.fn()
      const store = createStore({
        modules: {
          fields: {
            actions: {
              setModalFieldValue,
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })

      const input = wrapper.find( `input[name = "${ wrapper.vm.wpId }Left"]` )
      input.trigger( 'input' )
      await flushPromises()

      expect(setModalFieldValue).toBeCalledWith(
        expect.anything(),
        expect.objectContaining({
          name    : wrapper.vm.wpId,
          position: 'left',
          value   : val,
          as      : '',
          row     : '',
        }),
        void 0
      )
    })

    test('should set value if filled (row ver.)', async () => {
      const setModalFieldValue = jest.fn()
      const store = createStore({
        modules: {
          fields: {
            actions: {
              setModalFieldValue,
            }
          }
        }
      })
      const propsData = {
        as : 'row',
        row: 'testRow',
      }
      const wrapper = createWrapper({
        store,
        propsData,
      })

      const input = wrapper.find( `input[name = "${ wrapper.vm.wpId }Left"]` )
      input.trigger( 'input' )
      await flushPromises()

      expect(setModalFieldValue).toBeCalledWith(
        expect.anything(),
        expect.objectContaining({
          name    : wrapper.vm.wpId,
          position: 'left',
          value   : val,
          as      : propsData.as,
          row     : propsData.row,
        }),
        void 0
      )
    })

    test('should set value if filled (grid ver.)', async () => {
      const setModalFieldValue = jest.fn()
      const store = createStore({
        modules: {
          fields: {
            actions: {
              setModalFieldValue,
            }
          }
        }
      })
      const propsData = {
        as : 'grid',
      }
      const wrapper = createWrapper({
        store,
        propsData,
      })

      const input = wrapper.find( `input[name = "${ wrapper.vm.wpId }Left"]` )
      input.trigger( 'input' )
      await flushPromises()

      expect(setModalFieldValue).toBeCalledWith(
        expect.anything(),
        expect.objectContaining({
          name    : wrapper.vm.wpId,
          position: 'left',
          value   : val,
          as      : propsData.as,
          row     : '',
        }),
        void 0
      )
    })

  })

})