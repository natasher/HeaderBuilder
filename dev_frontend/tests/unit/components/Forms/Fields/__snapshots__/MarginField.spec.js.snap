// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField.vue snapshots: should match snapshot if located in grid form 1`] = `
<div
  class="row"
>
  <div
    class="col col-left"
  >
    <span
      class="title"
    >
      gridMargin
    </span>
  </div>
   
  <div
    class="col col-right"
  >
    <div
      class="mfn-field mfnf-margin"
    >
      <div
        class="container"
      >
        <input
          class="top"
          name="testFieldTop"
          placeholder="px"
          type="text"
        />
         
        <input
          class="right"
          name="testFieldRight"
          placeholder="px"
          type="text"
        />
         
        <input
          class="bottom"
          name="testFieldBottom"
          placeholder="px"
          type="text"
        />
         
        <input
          class="left"
          name="testFieldLeft"
          placeholder="px"
          type="text"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`TextField.vue snapshots: should match snapshot if located in item form 1`] = `
<div
  class="row"
>
  <div
    class="col col-left"
  >
    <span
      class="title"
    >
      Test Field
    </span>
  </div>
   
  <div
    class="col col-right"
  >
    <div
      class="mfn-field mfnf-margin"
    >
      <div
        class="container"
      >
        <input
          class="top"
          name="testFieldTop"
          placeholder="px"
          type="text"
        />
         
        <input
          class="right"
          name="testFieldRight"
          placeholder="px"
          type="text"
        />
         
        <input
          class="bottom"
          name="testFieldBottom"
          placeholder="px"
          type="text"
        />
         
        <input
          class="left"
          name="testFieldLeft"
          placeholder="px"
          type="text"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`TextField.vue snapshots: should match snapshot if located in row form 1`] = `
<div
  class="row"
>
  <div
    class="col col-left"
  >
    <span
      class="title"
    >
      Test Field
    </span>
  </div>
   
  <div
    class="col col-right"
  >
    <div
      class="mfn-field mfnf-margin"
    >
      <div
        class="container"
      >
        <input
          class="top"
          name="testFieldTop"
          placeholder="px"
          type="text"
        />
         
        <input
          class="right"
          name="testFieldRight"
          placeholder="px"
          type="text"
        />
         
        <input
          class="bottom"
          name="testFieldBottom"
          placeholder="px"
          type="text"
        />
         
        <input
          class="left"
          name="testFieldLeft"
          placeholder="px"
          type="text"
        />
      </div>
    </div>
  </div>
</div>
`;
